Allow paths to be hidden/shown in wxFileHistory menus

imciner2 opened the pull request:
This PR adds the ability to customize the menu item labels in menus associated with a file history, so that one of three behaviors is done:
* The path is hidden if it matches the first file in the list (this is the current behavior, and remains the default if it is not set explicitly)
* All paths can be hidden and only the filename + extension is shown
* All paths are shown

Example with no paths shown:
![NoPaths](https://user-images.githubusercontent.com/2262453/111563489-fa0a8a80-878f-11eb-92a5-97c7f452fc07.png)

Customizing this is useful for an application because there doesn't appear to be a standard way of displaying a file history menu, for instance a survey of programs shows the following:
* LibreOffice - Just file name
* Pluma - Just file name
* Atril - Just file name
* FreeCAD - Just file name
* FoxitReader - Just file name
* Sublime Text - Full path
* VLC - Full path
* TexStudio - Full path
* wxFormBuilder - Full path
* Audacity - Full path

imciner2 committed:
Add a file history menu to the menu sample

imciner2 committed:
Add different menu label styles to wxFileHistory menus

These styles allow the program to choose to display file paths
in the menu or not

vadz commented on include/wx/filehistory.h:
```diff
@@ -24,6 +24,13 @@ class WXDLLIMPEXP_FWD_CORE wxMenu;
 
+enum wxFileHistoryMenuLabelStyle
+{
+    wxFH_HIDE_CURRENT_PATH = 0,
```
I'm not quite sure if "current" is very clear here. Maybe `wxFH_HIDE_SAME_PATH`? Or `wxFH_SHOW_NON_UNIQUE_PATH`?

vadz replied:
I've finally used `wxFH_PATH_SHOW_XXX` names for all of them.

vadz commented on include/wx/filehistory.h:
```diff
@@ -60,15 +67,23 @@ class WXDLLIMPEXP_CORE wxFileHistoryBase : public wxObject
+    void SetMenuLabelStyle(wxFileHistoryMenuLabelStyle style) { m_menuLabelStyle = style; }
+    wxFileHistoryMenuLabelStyle GetMenuLabelStyle() const { return m_menuLabelStyle; }
+
+    void RefreshLabels();
```
Do we really need to make this one public, i.e. is there any point in calling it from elsewhere than from `SetMenuLabelStyle()`? I would rather refresh the labels automatically when the label style changes, why not? I.e. the usual rationale for not updating automatically is that you can have a series of `SetXXX()` calls and want to have only a single `Refresh()` at the end, but here this doesn't risk happening because there is just a single setter.

vadz commented on include/wx/filehistory.h:
```diff
@@ -24,6 +24,13 @@ class WXDLLIMPEXP_FWD_CORE wxMenu;
 class WXDLLIMPEXP_FWD_BASE wxConfigBase;
 class WXDLLIMPEXP_FWD_BASE wxFileName;
 
+enum wxFileHistoryMenuLabelStyle
```
Sorry for the first (but not last) of naming nitpicks, but I think it would be useful to use the word "path" here, and in the names of the functions dealing with this enum, because otherwise it's not obvious at all that it has something to do with them. Maybe just call it `wxFileHistoryMenuPathStyle`?

vadz replied:
Just FYI, I've renamed it now.

vadz commented on include/wx/filehistory.h:
```diff
@@ -60,15 +67,23 @@ class WXDLLIMPEXP_CORE wxFileHistoryBase : public wxObject
     void SetBaseId(wxWindowID baseId) { m_idBase = baseId; }
     wxWindowID GetBaseId() const { return m_idBase; }
 
+    void SetMenuLabelStyle(wxFileHistoryMenuLabelStyle style) { m_menuLabelStyle = style; }
+    wxFileHistoryMenuLabelStyle GetMenuLabelStyle() const { return m_menuLabelStyle; }
```
We'd need rename those to `{Set,Get}MenuPathStyle()` too if we rename the enum.

vadz commented on interface/wx/filehistory.h:
```diff
@@ -5,6 +5,23 @@
 /////////////////////////////////////////////////////////////////////////////
 
+/**
+ Styles for the labels on menu items for wxFileHistory menus
+
+ @since 3.1.5
```
Very very minor, but the indentation should be more than one space here.

vadz commented on interface/wx/filehistory.h:
```diff
@@ -5,6 +5,23 @@
+*/
+enum wxFileHistoryMenuLabelStyle
+{
+    /** Hide the file path if it matches the path of the first item */
```
I would document that this is the default value here (too).

vadz commented on interface/wx/filehistory.h:
```diff
@@ -15,6 +32,8 @@
     required in an MDI application, where the file history should appear on
     each MDI child menu as well as the MDI parent frame.
 
+    By default, the menu item label style will be @c wxFH_HIDE_CURRENT_PATH.
```
```suggestion
    By default, the menu item label style is @c wxFH_HIDE_CURRENT_PATH,
    meaning that paths are only shown in the menu items if they differ.
    This can be changed using SetMenuLabelStyle().
```

vadz commented on src/common/filehistorycmn.cpp:
```diff
@@ -135,20 +136,43 @@ void wxFileHistoryBase::AddFileToHistory(const wxString& file)
+
+void wxFileHistoryBase::RefreshLabels()
+{
+    size_t i;
```
Very minor, but could this be declared inside the loop which uses it instead?

vadz commented on src/common/filehistorycmn.cpp:
```diff
@@ -135,20 +136,43 @@ void wxFileHistoryBase::AddFileToHistory(const wxString& file)
+void wxFileHistoryBase::RefreshLabels()
+{
+    size_t i;
+    size_t numFiles = m_fileHistory.size();
```
Another pretty minor improvement
```suggestion
    const size_t numFiles = m_fileHistory.size();
```

vadz commented on src/common/filehistorycmn.cpp:
```diff
@@ -135,20 +136,43 @@ void wxFileHistoryBase::AddFileToHistory(const wxString& file)
+        return;
+
+    // Use the first file as the current path
+    wxFileName firstFn(m_fileHistory[0]);
```
It looks like it could be simpler if we just did
```suggestion
    const wxString firstPath = wxFileName(m_fileHistory[0]).GetPath();
```
and then compared the path with it below.

vadz reviewed (commented):
Thanks! I agree that this is useful to have, but I have doubts about naming (and exposing/requiring to call `RefreshLabels()`). Please let me know if you think any of my suggestions have merit or if you disagree.

vadz commented:
I'll try to make the necessary changes, i.e. notably get rid of `RefreshLabels()`, before 3.1.5. Please let me know if you have any objections or, alternatively, you're perhaps already working on changing this to avoid work duplication. TIA!

imciner2 commented:
> I'll try to make the necessary changes, i.e. notably get rid of `RefreshLabels()`, before 3.1.5. Please let me know if you have any objections or, alternatively, you're perhaps already working on changing this to avoid work duplication. TIA!

Thanks, I hadn't had a chance to get back to this because I had been chasing down some other bugs/problems in a downstream application - so you can go ahead and make the changes without duplicating work.

vadz committed:
Rename wxFileHistoryMenuLabelStyle to wxFileHistoryMenuPathStyle

It seems useful to have the word "Path" in the name of this enum to
indicate that it applies to the paths shown in the menu labels.

Also rename the methods using this enum.

vadz committed:
Rename wxFileHistoryMenuPathStyle values too

Use common wxFH_PATH_ prefix for consistency and try to make the values
more clear.

vadz committed:
Remove public wxFileHistory::RefreshLabels()

Replace it with a private DoRefreshLabels() and call it ourselves from
SetMenuPathStyle() to make the class simpler (and less error-prone, as
it's now impossible to forget to call RefreshLabels() any more) to use.

vadz committed:
Very minor code cleanup in AddFileToHistory()

Declare "i" inside the loop and make a variable that could be const
actually const.

No real changes.

vadz committed:
Make DoRefreshLabels() code more obviously correct

Use switch over enum instead of a series of chained ifs.

This has the same effect, but can be read and understood more easily.

No real changes.

vadz committed:
Fix wxFileHistory memory leak in the sample

We need to delete the object we allocate.

vadz commented:
I've done all the changes I had in mind, please let me know if you have any comments, otherwise I'll merge this soon. TIA!
